### DISCUSSION TITLE
TypeError: String.__init__() got an unexpected keyword argument '_enums'

### DISCUSSION THREAD

**Discussion by chalmerlowe** [sqlalchemy/sqlalchemy#11734] [Usage Questions]
I am attempting to run the compliances tests with the BigQuery Dialect (I am the Python-BigQuery maintainer).

When attempting [to execute the tests in `EnumTest`](https://github.com/sqlalchemy/sqlalchemy/blob/fdac33149b85480b7ec34e86f031afb8a0a9e54f/lib/sqlalchemy/testing/suite/test_types.py#L1922):
* `EnumTest::test_round_trip_executemany`
* `EnumTest::test_round_trip_executemany_returning`
* `EnumTest::test_round_trip["a"]`
* `EnumTest::test_round_trip["b"]`
* `EnumTest::test_round_trip["a%"]`
* `EnumTest::test_round_trip["b%percent"]`
* `EnumTest::test_round_trip["réveillé"]`

We get this error: 
```
TypeError: String.__init__() got an unexpected keyword argument '_enums'
```

In this case, `String` is [defined in sqlalchemy, here](https://github.com/sqlalchemy/sqlalchemy/blob/fdac33149b85480b7ec34e86f031afb8a0a9e54f/lib/sqlalchemy/sql/sqltypes.py#L169).

As the test code progresses through various machinations, a keyword argument `kw` is created and it appears to be carried through the entire process until it gets to the initialization process for `String`, at which time, `String.__init__()` doesn't know what to do with it.

In attempting to trace the code, I find that sqlalchemy includes in `kw` an `Enum` object using the keyword `"_enums"` when trying to call [`constructor_copy()`](https://github.com/sqlalchemy/sqlalchemy/blob/fdac33149b85480b7ec34e86f031afb8a0a9e54f/lib/sqlalchemy/util/langhelpers.py#L1402) in `sqlalchemy/util/langhelpers.py", which attempts to create another object.

This ultimately leads to an attempt to create a `BQString` object. `BQString` inherits from sqlalchemy's [`String` object](https://github.com/sqlalchemy/sqlalchemy/blob/fdac33149b85480b7ec34e86f031afb8a0a9e54f/lib/sqlalchemy/sql/sqltypes.py#L169), which initializes with just `length` and `collation`.

```
def __init__(
        self,
        length: Optional[int] = None,
        collation: Optional[str] = None,
    ):
```

Thus when we attempt to initialize a `String` object with a `kw` argument that includes content beyond `length` and `collation`, we get the above error.

The function `constructor_copy` has code that is supposed to pare down the keywords, but it appears that it does not do it in all circumstances (ie. when `kw` comes to the table with pre-existing keywords).

```
names = get_cls_kwargs(cls)
kw.update(
    (k, obj.__dict__[k]) for k in names.difference(kw) if k in obj.__dict__
)
return cls(*args, **kw)
```

Any help you can offer, if I am missing something, would be appreciated. If this is some sort of bug, I am happy to issue a Bug.

**Comment by zzzeek** [comment]
does BigQuery have an ENUM type of its own ?   or are you assuming SQLAlchemy's Enum will always use "non native" mode ?

**Reply by zzzeek** [reply]
additionally what is BQString used for and how is it registered

**Reply by chalmerlowe** [reply]
BigQuery (BQ) does not have an Enum datatype.
BQString has one primary job: to ensure that the string doesn't have character sequences that are not allowed by BQ.

**Comment by zzzeek** [comment]
Assuming your BQString is registered in colspecs, I can reproduce the problem here.  The problem is that Enum is a subclass of String, so if you are defining a new dialect specific class of String, you would need to accommodate for Enum as well.  The simplest way assuming you dont need BQString's logic for Enum is to map Enum to itself, which would bypass the BQString type for Enum.  Here's what it looks like in the SQLite dialect:

```diff
diff --git a/lib/sqlalchemy/dialects/sqlite/base.py b/lib/sqlalchemy/dialects/sqlite/base.py
index 04e84a68d..b5424e5a1 100644
--- a/lib/sqlalchemy/dialects/sqlite/base.py
+++ b/lib/sqlalchemy/dialects/sqlite/base.py
@@ -1241,6 +1241,8 @@ class TIME(_DateTimeMixin, sqltypes.Time):
         else:
             return processors.str_to_time
 
+class BQString(sqltypes.String):
+    pass
 
 colspecs = {
     sqltypes.Date: DATE,
@@ -1249,6 +1251,8 @@ colspecs = {
     sqltypes.JSON.JSONIndexType: JSONIndexType,
     sqltypes.JSON.JSONPathType: JSONPathType,
     sqltypes.Time: TIME,
+    sqltypes.String: BQString,
+    sqltypes.Enum: sqltypes.Enum,
 }
 
 ischema_names = {

```

the key issue here is that it is very unusual to define a dialect-specific class for `String`, as this would imply BigQuery has special steps required before string values can be passed to or received from the  DBAPI used by BigQuery, which should not be the case in Python 3- the underlying DBAPI (which is not the same thing as the dialect) should accept and return Python Unicode strings directly to the client for all VARCHAR  / CHAR types of columns in the database.

**Comment by chalmerlowe** [comment]
I updated our dialect to handle Enums in a manner similar to what is done in SQLite.
Thanks for you time and information. It appears that this has resolved my issue.
